File: selection/harness.rb
# Generic Ruby loop-harness for the cross-language transfer check.
# Reads the SAME shared JSON inputs as the Python/JS harnesses, loads the cell's
# solution.rb, and loops the call to a wall-time budget (checksum-consumed,
# iteration-counted). Ruby is dynamic, so plain-data args pass straight through;
# only tree/list/design need construction.
#
#   ruby harness.rb <budget_seconds> <case_index>   (run from the Ruby cell dir)
#   - slug derived from the cell directory name (Ruby/leetcode/<slug>/)
#   - inputs from ../../../reference/leetcode/{outputs,workloads}/<slug>.json
#   - prints CASE/ITERS/ACC/BEACON to STDERR (stdout stays empty -> validate ok)
#
# LeetCode Ruby methods are snake_case while the shared entry_point / design ops
# are camelCase, so method names are converted (Rails-style "underscore") with a
# respond_to? fallback to the original spelling.
require 'json'
require 'set' # LeetCode's Ruby runtime preloads stdlib 'set'; match that environment

# ---- node types the solutions expect (they only ever receive these) ----------
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(v); @val = v; end
end
class ListNode
  attr_accessor :val, :next
  def initialize(v); @val = v; end
end

module HNS
  module_function

  def underscore(s)
    s = s.to_s.dup
    s.gsub!(/([A-Z]+)([A-Z][a-z])/, '\1_\2') # acronym boundary: MKAverage -> MK_Average
    s.gsub!(/([a-z\d])([A-Z])/, '\1_\2')      # camel boundary:   addElement -> add_Element
    s.downcase
  end

  def build_tree(a)
    return nil if a.nil? || a.empty? || a[0].nil?
    root = TreeNode.new(a[0]); q = [root]; i = 1
    until i >= a.length || q.empty?
      n = q.shift
      if i < a.length
        lv = a[i]; i += 1
        if !lv.nil?; n.left = TreeNode.new(lv); q << n.left; end
      end
      if i < a.length
        rv = a[i]; i += 1
        if !rv.nil?; n.right = TreeNode.new(rv); q << n.right; end
      end
    end
    root
  end

  def build_list(a)
    return nil if a.nil? || a.empty?
    head = ListNode.new(a[0]); cur = head
    a[1..-1].each { |v| cur.next = ListNode.new(v); cur = cur.next }
    head
  end

  def tree_to_arr(r)
    a = []; q = [r]
    until q.empty?
      n = q.shift
      if n.nil?; a << nil
      else; a << n.val; q << n.left; q << n.right; end
    end
    a.pop while !a.empty? && a[-1].nil?
    a
  end

  def list_to_arr(h)
    a = []
    while h; a << h.val; h = h.next; end
    a
  end

  def canon(r)
    case r
    when nil     then 'null'
    when TreeNode then tree_to_arr(r).to_json
    when ListNode then list_to_arr(r).to_json
    else
      begin; r.to_json; rescue StandardError; r.to_s; end
    end
  end

  MASK64 = (1 << 64) - 1
  def strhash(s)
    h = 0
    s.each_byte { |b| h = (h * 31 + b) & 0xFFFFFFFF }
    h
  end
  def fold(acc, r)
    (acc * 1000003 + strhash(canon(r))) & MASK64
  end

  # deep copy of plain data (numbers/strings/arrays/hashes/bools)
  def deep(x)
    Marshal.load(Marshal.dump(x))
  end

  # pick the actual Ruby method name (snake_case first, camel fallback)
  def resolve(obj, camel)
    snake = underscore(camel)
    return snake if obj.respond_to?(snake, true)
    return camel if obj.respond_to?(camel, true)
    snake
  end
end

# --------------------------------------------------------------------------- #
budget = Float(ARGV[0])
idx    = Integer(ARGV[1])
cell   = Dir.pwd
slug   = File.basename(cell)
ref    = File.join(cell, '..', '..', '..', 'reference', 'leetcode')

out   = JSON.parse(File.read(File.join(ref, 'outputs', "#{slug}.json")))
kase  = out['expected'][idx]
name  = kase['name']
input = kase['input']

# load the solution (top-level defs become private Object methods; classes -> constants)
prev_verbose = $VERBOSE
$VERBOSE = nil
load File.join(cell, 'solution.rb')
$VERBOSE = prev_verbose

warm = budget * 0.3
def now; Process.clock_gettime(Process::CLOCK_MONOTONIC); end

acc = 0
iters = 0
echo = ENV['HNS_ECHO']

if input.is_a?(Hash) && input.key?('ops') && input.key?('args')
  # ---- design-class replay: ops[0]=class, args[0]=ctor args, then method calls ----
  ops  = input['ops']
  args = input['args']
  klass = begin
            Object.const_get(ops[0])
          rescue NameError
            Object.const_get('Solution') # solution named the class differently
          end
  replay = lambda do |collect|
    inst = klass.new(*HNS.deep(args[0]))
    a = 0
    results = collect ? [nil] : nil
    (1...ops.length).each do |i|
      m = HNS.resolve(inst, ops[i])
      r = inst.send(m, *HNS.deep(args[i]))
      results << r if collect
      a = HNS.fold(a, r)
    end
    collect ? results : a
  end

  if echo
    STDERR.puts "RESULT=#{replay.call(true).to_json}"
  end
  # stateful replay -> args deep-copied each pass (kept); batch the clock read.
  wi = 0; t0 = now
  while now - t0 < warm; replay.call(false); wi += 1; end
  per = wi > 0 ? (now - t0) / wi : warm
  batch = per > 0 ? [[1, (0.002 / per).to_i].max, 4096].min : 4096
  tm = now
  while now - tm < budget - warm
    batch.times { acc = HNS.fold(acc, replay.call(false)) }
    iters += batch
  end
  meas = now - tm
  STDERR.puts "CASE=#{name} ITERS=#{iters} ACC=#{acc} MEAS_S=#{'%.6f' % meas} BEACON=design"
else
  # ---- generic single-method call ----
  wl     = JSON.parse(File.read(File.join(ref, 'workloads', "#{slug}.json")))
  ep     = wl['entry_point'].to_s
  camel  = ep.split('.').last
  keys   = input.is_a?(Hash) ? input.keys : []

  has_sol = defined?(Solution) && Solution.is_a?(Class)
  recv    = has_sol ? Solution.new : Object.new
  meth    = HNS.resolve(recv, camel)

  build_args = lambda do
    keys.map do |k|
      v = input[k]
      if k == 'root'    then HNS.build_tree(HNS.deep(v))
      elsif k == 'head' then HNS.build_list(HNS.deep(v))
      else HNS.deep(v)
      end
    end
  end
  # (1) detect mutation once; reuse deep-copied args when the solution leaves
  #     them untouched (HNS.deep per iteration was the dominant Ruby overhead).
  pristine = build_args.call
  snap = pristine.map { |x| HNS.canon(x) }.join("\x1e")
  probe = build_args.call
  beacon = recv.send(meth, *probe)
  mutated = probe.map { |x| HNS.canon(x) }.join("\x1e") != snap
  one = mutated ? lambda { recv.send(meth, *build_args.call) } : lambda { recv.send(meth, *pristine) }

  STDERR.puts "RESULT=#{HNS.canon(beacon)}" if echo
  # (3) estimate per-iter cost during warmup, then batch the wall-clock read
  wi = 0; t0 = now
  while now - t0 < warm; one.call; wi += 1; end
  per = wi > 0 ? (now - t0) / wi : warm
  batch = per > 0 ? [[1, (0.002 / per).to_i].max, 4096].min : 4096
  tm = now
  while now - tm < budget - warm
    batch.times { acc = HNS.fold(acc, one.call) }
    iters += batch
  end
  meas = now - tm
  STDERR.puts "CASE=#{name} ITERS=#{iters} ACC=#{acc} MEAS_S=#{'%.6f' % meas} BEACON=#{HNS.canon(beacon)[0, 80]}"
end

File: selection/harness.js
// Generic JavaScript loop-harness for the cross-language transfer check.
// Reads the SAME shared JSON inputs, calls the solution, loops to a wall-time
// budget (checksum-consumed, iteration-counted). JS is dynamic, so plain-data
// args pass straight through; only tree/list/design need construction.
//
//   node harness.js <budget_seconds> <case_index>   (run from the JS cell dir)
//   prints CASE/ITERS/ACC/BEACON to STDERR (stdout stays empty).
const fs = require('fs');
const path = require('path');

function TreeNode(v){ this.val=v; this.left=null; this.right=null; }
function ListNode(v){ this.val=v; this.next=null; }
function buildTree(a){
  if(!a || !a.length || a[0]==null) return null;
  const root=new TreeNode(a[0]); const q=[root]; let i=1;
  while(i<a.length && q.length){ const n=q.shift();
    if(i<a.length){ const lv=a[i++]; if(lv!=null){ n.left=new TreeNode(lv); q.push(n.left); } }
    if(i<a.length){ const rv=a[i++]; if(rv!=null){ n.right=new TreeNode(rv); q.push(n.right); } } }
  return root;
}
function buildList(a){ if(!a || !a.length) return null; const h=new ListNode(a[0]); let c=h;
  for(let i=1;i<a.length;i++){ c.next=new ListNode(a[i]); c=c.next; } return h; }
function clone(o){
  if(Array.isArray(o)) return o.map(clone);
  if(o instanceof TreeNode) return buildTree(treeToArr(o));
  if(o instanceof ListNode) return buildList(listToArr(o));
  return o;                                   // primitives / strings
}
function treeToArr(r){ const a=[]; const q=[r]; while(q.length){ const n=q.shift();
    if(n==null){ a.push(null); } else { a.push(n.val); q.push(n.left); q.push(n.right); } }
  while(a.length && a[a.length-1]==null) a.pop(); return a; }
function listToArr(h){ const a=[]; while(h){ a.push(h.val); h=h.next; } return a; }
function canon(r){
  if(r===undefined || r===null) return "null";          // void design methods return undefined
  if(r instanceof ListNode) return JSON.stringify(listToArr(r));
  if(r instanceof TreeNode) return JSON.stringify(treeToArr(r));
  try { const s = JSON.stringify(r); return s === undefined ? "null" : s; } catch(e){ return String(r); }
}
// V8's native JSON.stringify + a simple int charloop is faster than a pure-JS
// recursive structural checksum (per-node type dispatch dominates), so the fold
// stays string-based here — the win in JS comes from skipping the input clone.
function fold(acc, r){ let h=0; const s=canon(r); for(let i=0;i<s.length;i++) h=(h*31 + s.charCodeAt(i))|0; return (acc*1000003 + h)|0; }

const cell = process.cwd();
const slug = path.basename(cell);
const ref  = path.join(cell, '..', '..', '..', 'reference', 'leetcode');
const out  = JSON.parse(fs.readFileSync(path.join(ref,'outputs',slug+'.json'),'utf8'));
const budget = parseFloat(process.argv[2]);
const idx    = parseInt(process.argv[3]);
const kase = out.expected[idx];
const input = kase.input;

// bring the solution's top-level defs into this scope
eval(fs.readFileSync(path.join(cell,'solution.js'),'utf8'));

function now(){ return process.hrtime.bigint(); }
function secs(a,b){ return Number(b-a)/1e9; }
let acc=0, iters=0, beacon;

if('ops' in input && 'args' in input){
  // design-class replay: ops[0] = class name, args[0] = ctor args, then method calls
  const ops=input.ops, args=input.args;
  const Cls = eval(ops[0]);
  const replay = () => { const inst = new Cls(...clone(args[0])); let a=0;
    for(let i=1;i<ops.length;i++) a=fold(a, inst[ops[i]](...clone(args[i]))); return a; };
  // stateful replay -> args cloned each pass (kept); batch the clock read.
  let wi=0, tw=now(); while(secs(tw,now()) < budget*0.3){ replay(); wi++; }
  const perIter = wi ? secs(tw,now())/wi : budget*0.3;
  const batch = perIter>0 ? Math.max(1, Math.min(4096, Math.floor(0.002/perIter))) : 4096;
  let tm=now();
  while(secs(tm,now()) < budget*0.7){ for(let b=0;b<batch;b++){ acc=(acc*1000003 + replay())|0; } iters+=batch; }
  const meas=secs(tm,now());
  process.stderr.write(`CASE=${kase.name} ITERS=${iters} ACC=${acc} MEAS_S=${meas.toFixed(6)} BEACON=design\n`);
} else {
  const wl = JSON.parse(fs.readFileSync(path.join(ref,'workloads',slug+'.json'),'utf8'));
  const method = String(wl.entry_point||'').split('.').pop();
  // build args (tree/list by conventional key name; else pass through)
  const base = Object.keys(input).map(k =>
      k==='root' ? buildTree(input[k]) : k==='head' ? buildList(input[k]) : input[k]);
  const hasSol = (typeof Solution !== 'undefined');
  const call = (a) => hasSol ? new Solution()[method](...a) : eval(method)(...a);
  // (1) detect mutation once; skip the per-iteration clone when the solution
  //     leaves its input untouched (the common case). `base` is a throwaway
  //     probe copy; `pristine` is the clean reference reused when non-mutating.
  const pristine = base.map(clone);
  const snap = pristine.map(canon).join('\x1e');
  beacon = call(base);
  const mutated = base.map(canon).join('\x1e') !== snap;
  const one = mutated ? (() => call(pristine.map(clone))) : (() => call(pristine));
  // (3) estimate per-iter cost during warmup, then batch the wall-clock read
  let wi=0, tw=now(); while(secs(tw,now()) < budget*0.3){ one(); wi++; }
  const perIter = wi ? secs(tw,now())/wi : budget*0.3;
  const batch = perIter>0 ? Math.max(1, Math.min(4096, Math.floor(0.002/perIter))) : 4096;
  let tm=now();
  while(secs(tm,now()) < budget*0.7){ for(let b=0;b<batch;b++){ acc=fold(acc, one()); } iters+=batch; }
  const meas=secs(tm,now());
  process.stderr.write(`CASE=${kase.name} ITERS=${iters} ACC=${acc} MEAS_S=${meas.toFixed(6)} BEACON=${canon(beacon).slice(0,80)}\n`);
}
